# Activity: Explore descriptive statistics

## **Introduction**

Data professionals often use descriptive statistics to understand the data they are working with and provide collaborators with a summary of the relative location of values in the data, as well an information about its spread. 

For this activity, you are a member of an analytics team for the United States Environmental Protection Agency (EPA). You are assigned to analyze data on air quality with respect to carbon monoxide, a major air pollutant. The data includes information from more than 200 sites, identified by state, county, city, and local site names. You will use Python functions to gather statistics about air quality, then share insights with stakeholders.

## **Step 1: Imports** 


Import the relevant Python libraries `pandas` and `numpy`.

In [2]:
# Import relevant Python libraries.

import pandas as pd
import numpy as np


The dataset provided is in the form of a .csv file named `c4_epa_air_quality.csv`. It contains a susbet of data from the U.S. EPA. As shown in this cell, the dataset has been automatically loaded in for you. You do not need to download the .csv file, or provide more code, in order to access the dataset and proceed with this lab. Please continue with this activity by completing the following instructions.

In [3]:
# RUN THIS CELL TO IMPORT YOUR DATA.

air_quality = pd.read_csv('c4_epa_air_quality.csv', index_col=0)



<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

  There is a function in the `pandas` library that allows you to read in data from a .csv file and load it into a DataFrame. 

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  Use the `read_csv` function from the pandas `library`. The `index_col` parameter can be set to `0` to read in the first column as an index (and to avoid `"Unnamed: 0"` appearing as a column in the resulting DataFrame).

</details>

## **Step 2: Data exploration** 

To understand how the dataset is structured, display the first 10 rows of the data.

In [4]:
# Display first 10 rows of the data.

air_quality.head(10)

,date_local,state_name,county_name,city_name,local_site_name,parameter_name,units_of_measure,arithmetic_mean,aqi
0,2018-01-01,Arizona,Maricopa,Buckeye,BUCKEYE,Carbon monoxide,Parts per million,0.473684,7
1,2018-01-01,Ohio,Belmont,Shadyside,Shadyside,Carbon monoxide,Parts per million,0.263158,5
2,2018-01-01,Wyoming,Teton,Not in a city,Yellowstone National Park - Old Faithful Snow ...,Carbon monoxide,Parts per million,0.111111,2
3,2018-01-01,Pennsylvania,Philadelphia,Philadelphia,North East Waste (NEW),Carbon monoxide,Parts per million,0.300000,3
4,2018-01-01,Iowa,Polk,Des Moines,CARPENTER,Carbon monoxide,Parts per million,0.215789,3
5,2018-01-01,Hawaii,Honolulu,Not in a city,Kapolei,Carbon monoxide,Parts per million,0.994737,14
6,2018-01-01,Hawaii,Honolulu,Not in a city,Kapolei,Carbon monoxide,Parts per million,0.200000,2
7,2018-01-01,Pennsylvania,Erie,Erie,NaN,Carbon monoxide,Parts per million,0.200000,2
8,2018-01-01,Hawaii,Honolulu,Honolulu,Honolulu,Carbon monoxide,Parts per million,0.400000,5
9,2018-01-01,Colorado,Larimer,Fort Collins,Fort Collins - CSU - S. Mason,Carbon monoxide,Parts per million,0.300000,6


<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

  There is a function in the `pandas` library that allows you to get a specific number of rows from the top of a DataFrame. 

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  Use the `head()` function from the `pandas` library.

</details>

**Question:** What does the `aqi` column represent?

A: Air Quality Index

**Question:** In what units are the aqi values expressed?

A: No units. It's an index.

Now, get a table that contains some descriptive statistics about the data.

In [7]:
# Get descriptive stats.

print(air_quality.describe(include='all'))

        date_local  state_name  county_name      city_name local_site_name  \
count          260         260          260            260             257   
unique           1          52          149            190             253   
top     2018-01-01  California  Los Angeles  Not in a city         Kapolei   
freq           260          66           14             21               2   
mean           NaN         NaN          NaN            NaN             NaN   
std            NaN         NaN          NaN            NaN             NaN   
min            NaN         NaN          NaN            NaN             NaN   
25%            NaN         NaN          NaN            NaN             NaN   
50%            NaN         NaN          NaN            NaN             NaN   
75%            NaN         NaN          NaN            NaN             NaN   
max            NaN         NaN          NaN            NaN             NaN   

         parameter_name   units_of_measure  arithmetic_mean    

<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

  There is a function in the `pandas` library that allows you to generate a table of basic descriptive statistics about the numeric columns in a DataFrame.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  Use the `describe()` function from the `pandas` library.

</details>

**Question:** Based on the table of descriptive statistics, what do you notice about the count value for the `aqi` column?

A: There are 260 data points

**Question:** What do you notice about the 25th percentile for the `aqi` column?

This is an important measure for understanding where the aqi values lie. 

A: first quartile goes up to 2

**Question:** What do you notice about the 75th percentile for the `aqi` column?

This is another important measure for understanding where the aqi values lie. 

A: third quartile is at 9

## **Step 3: Statistical tests** 

Next, get some descriptive statistics about the states in the data.

In [9]:
# Get descriptive stats about the states in the data.
air_quality['state_name'].describe()

count            260
unique            52
top       California
freq              66
Name: state_name, dtype: object

<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

  There is a function in the `pandas` library that allows you to generate basic descriptive statistics about a DataFrame or a column you are interested in.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

 Use the `describe()` function from the `pandas` library. Note that this function can be used:
- "on a DataFrame (to find descriptive statistics about the numeric columns)" 
- "directly on a column containing categorical data (to find pertinent descriptive statistics)"

</details>

**Question:** What do you notice while reviewing the descriptive statistics about the states in the data? 

Note: Sometimes you have to individually calculate statistics. To review to that approach, use the `numpy` library to calculate each of the main statistics in the preceding table for the `aqi` column.

A: Gives the information from the states data that is NOT numerical (calculated) 

## **Step 4. Results and evaluation**

Now, compute the mean value from the `aqi` column.

In [10]:
# Compute the mean value from the aqi column.

print(air_quality['aqi'].mean())



6.757692307692308


<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

  There is a function in the `numpy` library that allows you to get the mean value from an array or a Series of values.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  Use the `mean()` function from the `numpy` library.

</details>

**Question:** What do you notice about the mean value from the `aqi` column?

This is an important measure, as it tells you what the average air quality is based on the data.

A: Average air quality index in the data set is 6.76

Next, compute the median value from the aqi column.

In [11]:
# Compute the median value from the aqi column.

print(air_quality['aqi'].median())

5.0


<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

  There is a function in the `numpy` library that allows you to get the median value from an array or a series of values.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  Use the `median()` function from the `numpy` library.

</details>

**Question:** What do you notice about the median value from the `aqi` column?

This is an important measure for understanding the central location of the data.

A: Median value is 5 so the data centralizes around 5.

Next, identify the minimum value from the `aqi` column.

In [12]:
# Identify the minimum value from the aqi column.

print(air_quality['aqi'].min())

0


<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

Refer to the video about descriptive statistics in Python.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  There is a function in the `numpy` library that allows you to get the minimum value from an array or a Series of values.

</details>

<details>
  <summary><h4><strong>Hint 3</strong></h4></summary>

  Use the `min()` function from the `numpy` library.

</details>

**Question:** What do you notice about the minimum value from the `aqi` column?

This is an important measure, as it tell you the best air quality observed in the data.

The minimum value for the aqi column is 0. This means that the smallest aqi value in the data is 0 parts per million. It is the best air quality o the data set.


Now, identify the maximum value from the `aqi` column.

In [13]:
# Identify the maximum value from the aqi column.

print(air_quality['aqi'].max())

50


<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

  There is a function in the `numpy` library that allows you to get the maximum value from an array or a Series of values.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  Use the `max()` function from the `numpy` library.

</details>

**Question:** What do you notice about the maximum value from the `aqi` column?

This is an important measure, as it tells you which value in the data corresponds to the worst air quality observed in the data.

A: Maximum index in the data set is 50, the worst air quality.

Now, compute the standard deviation for the `aqi` column.

By default, the `numpy` library uses 0 as the Delta Degrees of Freedom, while `pandas` library uses 1. To get the same value for standard deviation using either library, specify the `ddof` parameter to 1 when calculating standard deviation.

In [16]:
# Compute the standard deviation for the aqi column.

import statistics

print(np.std(air_quality['aqi'], ddof=1))


7.061706678820724


<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

  There is a function in the `numpy` library that allows you to get the standard deviation from an array or a series of values.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

Use the `std()` function from the `numpy` library. Make sure to specify the `ddof` parameter as 1. To read more about this function,  refer to its documentation in the references section of this lab.

</details>

**Question:** What do you notice about the standard deviation for the `aqi` column? 

This is an important measure of how spread out the aqi values are.

A: standard deviation is 7.06 so most air quality values are between 0 and 14

## **Considerations**


**What are some key takeaways that you learned during this lab?**

A: I learned how to use the descriptive statistics functions to describe data from a dataset/csv file

**How would you present your findings from this lab to others? Consider the following relevant points noted by AirNow.gov as you respond:**
- "AQI values at or below 100 are generally thought of as satisfactory. When AQI values are above 100, air quality is considered to be unhealthy—at first for certain sensitive groups of people, then for everyone as AQI values increase."
- "An AQI of 100 for carbon monoxide corresponds to a level of 9 parts per million."

A: I would present a map of the U.S. with a color legend depicting which areas are the worst in the United States.

**What summary would you provide to stakeholders? Use the same information provided previously from AirNow.gov as you respond.**

A: I would summarize the data by providing the averages and central tendency while also highlighting that the highest value falls well within what is considered satisfactory by AirNow.gov

**References**

[Air Quality Index - A Guide to Air Quality and Your Health](https://www.airnow.gov/sites/default/files/2018-04/aqi_brochure_02_14_0.pdf). (2014,February)

[Numpy.Std — NumPy v1.23 Manual](https://numpy.org/doc/stable/reference/generated/numpy.std.html)

US EPA, OAR. (2014, 8 July).[*Air Data: Air Quality Data Collected at Outdoor Monitors Across the US*](https://www.epa.gov/outdoor-air-quality-data). 